### ISSUE TITLE
Missing calls to fsync() can leave systems with no valid partition table

### ISSUE THREAD

**Issue by DemiMarie** [rust-disk-partition-management/gptman#137]
Operating systems do not guarantee that writes reach disk in the same order in which they are made, unless `O_DIRECT|O_SYNC` is used or explicit calls to a flush function (such as `fsync()` on Unix-like systems or `FlushFileBuffers()` on Windows) are made and succeed.  Therefore, it is possible that both the writes to the primary and backup partition tables can partially succeed, leaving the system with no valid partition table and therefore unbootable.

The solution is to add the needed flushing where necessary.  User code cannot provide this flushing because no user code runs in the places where barriers are needed.

**Comment by cecton** [comment]
Hello and welcome 👋 

Sorry for the delay I lost track of time.

Regarding your concern: I understand the importance of ensuring durable writes for GPT integrity. However, the library only accepts a generic `impl std::io::Write`, and it's not feasible for it to enforce or perform system-level flushes like `fsync()` or `FlushFileBuffers()` internally. This is by design, and ensuring proper flushing (when using something like a `File`) is the responsibility of the caller.

There used to be a CLI binary as part of this project that might've handled flushing, but it has been removed and is no longer supported.

**Comment by DemiMarie** [comment]
You need to call `fsync()` _between write operations done by the library_.  In particular, using an arbitrary `std::io::Write` is not safe.

I think it would make more sense to provide platform-specific implementations, or at least an extension trait that callers must implement.

**Comment by DemiMarie** [comment]
> The user calling `fsync()` after they called `Gpt::write_into()` isn't sufficient? It's the only write method they need to call, everything is done in memory.

I don’t think so.  There need to be write barriers to ensure that the primary GPT is not overwritten before the backup GPT has been synced and visa versa.  I would `fsync` immediately after opening the device, but then `O_DIRECT|O_SYNC` might be okay.

**Comment by DemiMarie** [comment]
I think this makes sense.
